tests/host_reference/gauge_force_reference.cpp

```diff
@@ -277,57 +301,102 @@ int gf_neighborIndexFullLattice(size_t i, int dx[], const lattice_t &lat)
+{
+  su3_matrix prev_matrix, curr_matrix;
+
+  memset(&curr_matrix, 0, sizeof(curr_matrix));
+
+  curr_matrix.e[0][0].real = 1;
+  curr_matrix.e[1][1].real = 1;
+  curr_matrix.e[2][2].real = 1;
+
+  for (int j = 0; j < len; j++) {
+    int lnkdir;
+
+    prev_matrix = curr_matrix;
+    if (GOES_FORWARDS(path[j])) {
+      // dx[path[j]] +=1;
+      lnkdir = path[j];
+    } else {
+      dx[OPP_DIR(path[j])] -= 1;
+      lnkdir = OPP_DIR(path[j]);
+    }
+
+    int nbr_idx = gf_neighborIndexFullLattice(i, dx, lat);
+    su3_matrix *lnk = sitelink[lnkdir] + nbr_idx;
+
+    if (GOES_FORWARDS(path[j])) {
+      mult_su3_nn(&prev_matrix, lnk, &curr_matrix);
+    } else {
+      mult_su3_na(&prev_matrix, lnk, &curr_matrix);
+    }
+
+    if (GOES_FORWARDS(path[j])) {
+      dx[path[j]] += 1;
+    } else {
+      // we already subtract one in the code above
+    }
+  } // j
+
+  return curr_matrix;
+}
+
+
 // this function compute one path for all lattice sites
 template <typename su3_matrix, typename Float>
 static void compute_path_product(su3_matrix *staple, su3_matrix **sitelink, int *path, int len, Float loop_coeff,
                                  int dir, const lattice_t &lat)
 {
-  su3_matrix prev_matrix, curr_matrix, tmat;
+  su3_matrix curr_matrix, tmat;
   int dx[4];
 
   for (size_t i = 0; i < lat.volume; i++) {
     memset(dx, 0, sizeof(dx));
-    memset(&curr_matrix, 0, sizeof(curr_matrix));
-
-    curr_matrix.e[0][0].real = 1.0;
-    curr_matrix.e[1][1].real = 1.0;
-    curr_matrix.e[2][2].real = 1.0;
 
     dx[dir] = 1;
-    for (int j = 0; j < len; j++) {
-      int lnkdir;
-
-      prev_matrix = curr_matrix;
-      if (GOES_FORWARDS(path[j])) {
-        // dx[path[j]] +=1;
-        lnkdir = path[j];
-      } else {
-        dx[OPP_DIR(path[j])] -= 1;
-        lnkdir = OPP_DIR(path[j]);
-      }
-
-      int nbr_idx = gf_neighborIndexFullLattice(i, dx, lat);
-      su3_matrix *lnk = sitelink[lnkdir] + nbr_idx;
 
-      if (GOES_FORWARDS(path[j])) {
-        mult_su3_nn(&prev_matrix, lnk, &curr_matrix);
-      } else {
-        mult_su3_na(&prev_matrix, lnk, &curr_matrix);
-      }
-
-      if (GOES_FORWARDS(path[j])) {
-        dx[path[j]] += 1;
-      } else {
-        // we already substract one in the code above
-      }
-
-    } // j
+    curr_matrix = compute_gauge_path(sitelink, i, path, len, dx, lat);
 
     su3_adjoint(&curr_matrix, &tmat);
     scalar_mult_add_su3_matrix(staple + i, &tmat, loop_coeff, staple + i);
   } // i
 }
 
+template <typename su3_matrix>
+static dcomplex compute_loop_trace(su3_matrix **sitelink, int *path, int len, double loop_coeff, const lattice_t &lat)
+{
+  su3_matrix tmat;
+  dcomplex accum;
+  memset(&accum, 0, sizeof(accum));
+  int dx[4];
+
+  for (size_t i = 0; i < lat.volume; i++) {
+    memset(dx, 0, sizeof(dx));
+    tmat = compute_gauge_path(sitelink, i, path, len, dx, lat);
+    auto tr = trace_su3(&tmat);
+    accum.real += tr.real;
```

Comment:
could not use `CADD` and `CMUL` a few lines below?

Reply:
`CADD` is a good call, there's no flavor for multiplying by a `double` but I popped something in locally, will push momentarily.

Reply:
https://github.com/lattice/quda/commit/0eacdf5a4fa5ffd1631db7509707665b0b59f8b4